internal/rukpak/source/containers_image.go

```diff
@@ -0,0 +1,271 @@
+	if _, err := copy.Image(ctx, policyContext, layoutRef, dockerRef, &copy.Options{
+		SourceCtx: i.SourceContext,
+	}); err != nil {
+		return nil, fmt.Errorf("error copying image: %w", err)
+	}
+	l.Info("pulled image", "ref", imgRef.String(), "digest", canonicalRef.Digest().String())
+
+	//////////////////////////////////////////////////////
+	//
+	// Mount the image we just pulled
+	//
+	//////////////////////////////////////////////////////
+	if err := i.unpackImage(ctx, unpackPath, layoutRef); err != nil {
+		return nil, fmt.Errorf("error unpacking image: %w", err)
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Delete other images. They are no longer needed.
+	//
+	//////////////////////////////////////////////////////
+	if err := i.deleteOtherImages(bundle.Name, canonicalRef.Digest()); err != nil {
+		return nil, fmt.Errorf("error deleting old images: %w", err)
+	}
+
+	return successResult(bundle.Name, unpackPath, canonicalRef), nil
+}
+
+func successResult(bundleName, unpackPath string, canonicalRef reference.Canonical) *Result {
+	return &Result{
+		Bundle:         os.DirFS(unpackPath),
+		ResolvedSource: &BundleSource{Type: SourceTypeImage, Name: bundleName, Image: &ImageSource{Ref: canonicalRef.String()}},
+		State:          StateUnpacked,
+		Message:        fmt.Sprintf("unpacked %q successfully", canonicalRef),
+	}
+}
+
+func (i *ContainersImageRegistry) Cleanup(_ context.Context, bundle *BundleSource) error {
+	return os.RemoveAll(i.bundlePath(bundle.Name))
+}
+
+func (i *ContainersImageRegistry) bundlePath(bundleName string) string {
+	return filepath.Join(i.BaseCachePath, bundleName)
+}
+
+func (i *ContainersImageRegistry) unpackPath(bundleName string, digest digest.Digest) string {
+	return filepath.Join(i.bundlePath(bundleName), digest.String())
+}
+
+func resolveCanonicalRef(ctx context.Context, imgRef reference.Named, imageCtx *types.SystemContext) (reference.Canonical, error) {
+	if canonicalRef, ok := imgRef.(reference.Canonical); ok {
+		return canonicalRef, nil
+	}
+
+	srcRef, err := docker.NewReference(imgRef)
+	if err != nil {
+		return nil, NewUnrecoverable(fmt.Errorf("error creating reference: %w", err))
+	}
+
+	imgSrc, err := srcRef.NewImageSource(ctx, imageCtx)
+	if err != nil {
+		return nil, fmt.Errorf("error creating image source: %w", err)
+	}
+	defer imgSrc.Close()
+
+	imgManifestData, _, err := imgSrc.GetManifest(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error getting manifest: %w", err)
+	}
+	imgDigest, err := manifest.Digest(imgManifestData)
+	if err != nil {
+		return nil, fmt.Errorf("error getting digest of manifest: %w", err)
+	}
+	return reference.WithDigest(reference.TrimNamed(imgRef), imgDigest)
+}
+
+func (i *ContainersImageRegistry) unpackImage(ctx context.Context, unpackPath string, imageReference types.ImageReference) error {
+	img, err := imageReference.NewImage(ctx, i.SourceContext)
+	if err != nil {
+		return fmt.Errorf("error reading image: %w", err)
+	}
```

Comment:
```suggestion
	img, err := imageReference.NewImage(ctx, i.SourceContext)
	if err != nil {
		return fmt.Errorf("error reading image: %w", err)
	}
	defer func() {
		if err := img.Close(); err != nil {
			panic(err)
		}
	}()
```
Missing call to img.Close().

Reply:
Panic, you think? We could also pull a logger out of the context and log the error.

Reply:
Nobody is going to see that log.
I'd either panic or raise the error further up the stack.

Reply:
Panic it is. I expect this will never happen, but if it does, agree that it is likely something for a maintainer to look into.